File: temporal/ui/configurable_param_editor.py
from pathlib import Path
from typing import Any, Callable, Iterator, Optional, Type, TypeVar, cast

import gradio as gr
import numpy as np

from temporal.color import Color
from temporal.gradient import Gradient
from temporal.image_source import ImageSource
from temporal.meta.configurable import BoolParam, ColorParam, ConfigurableParam, EnumParam, FloatParam, GradientParam, ImageParam, ImageSourceParam, IntParam, NoiseParam, PathParam, PatternParam, StringParam
from temporal.noise import Noise
from temporal.pattern import Pattern
from temporal.ui import Callback, ReadData, UIThing, UpdateData, UpdateRequest, Widget
from temporal.ui.color_editor import ColorEditor
from temporal.ui.dropdown import Dropdown
from temporal.ui.gradient_editor import GradientEditor
from temporal.ui.gradio_widget import GradioWidget
from temporal.ui.image_source_editor import ImageSourceEditor
from temporal.ui.noise_editor import NoiseEditor
from temporal.ui.path_editor import PathEditor
from temporal.ui.pattern_editor import PatternEditor
from temporal.ui.radio import Radio


T = TypeVar("T")
U = TypeVar("U")


class ConfigurableParamEditor(Widget):
    def __init__(
        self,
        param: ConfigurableParam[T],
        value: Optional[T | Callable[[], T]] = None,
    ) -> None:
        super().__init__()

        def make_static_value(type: Type[U], param: ConfigurableParam[U]) -> U:
            return value if isinstance(value, type) else param.default

        def make_dynamic_value(type: Type[U], param: ConfigurableParam[U]) -> U | Callable[[], U]:
            return cast(Callable[[], U], value) if callable(value) else make_static_value(type, param)

        if isinstance(param, BoolParam):
            self._widget = GradioWidget(gr.Checkbox, label = self._format_label(param.name), value = make_dynamic_value(bool, param))

        elif isinstance(param, IntParam) and param.ui_type == "box":
            self._widget = GradioWidget(gr.Number, label = self._format_label(param.name), precision = 0, minimum = param.minimum, maximum = param.maximum, step = param.step, value = make_dynamic_value(int, param))

        elif isinstance(param, IntParam) and param.ui_type == "slider":
            self._widget = GradioWidget(gr.Slider, label = self._format_label(param.name), minimum = param.minimum if param.minimum is not None else 0, maximum = param.maximum if param.maximum is not None else 1, step = param.step, value = make_dynamic_value(int, param))

        elif isinstance(param, FloatParam) and param.ui_type == "box":
            self._widget = GradioWidget(gr.Number, label = self._format_label(param.name), minimum = param.minimum, maximum = param.maximum, step = param.step, value = make_dynamic_value(float, param))

        elif isinstance(param, FloatParam) and param.ui_type == "slider":
            self._widget = GradioWidget(gr.Slider, label = self._format_label(param.name), minimum = param.minimum if param.minimum is not None else 0, maximum = param.maximum if param.maximum is not None else 1, step = param.step, value = make_dynamic_value(float, param))

        elif isinstance(param, StringParam) and param.ui_type == "box":
            self._widget = GradioWidget(gr.Textbox, label = self._format_label(param.name), value = make_dynamic_value(str, param))

        elif isinstance(param, StringParam) and param.ui_type == "area":
            self._widget = GradioWidget(gr.TextArea, label = self._format_label(param.name), value = make_dynamic_value(str, param))

        elif isinstance(param, StringParam) and param.ui_type == "code":
            self._widget = GradioWidget(gr.Code, label = self._format_label(param.name), language = cast(None, param.language), value = make_dynamic_value(str, param))

        elif isinstance(param, PathParam):
            self._widget = PathEditor(label = self._format_label(param.name), value = make_dynamic_value(Path, param))

        elif isinstance(param, EnumParam) and param.ui_type == "menu":
            self._widget = Dropdown(label = self._format_label(param.name), choices = param.choices, value = make_dynamic_value(str, param))

        elif isinstance(param, EnumParam) and param.ui_type == "radio":
            self._widget = Radio(label = self._format_label(param.name), choices = param.choices, value = make_dynamic_value(str, param))

        elif isinstance(param, ColorParam):
            self._widget = ColorEditor(label = self._format_label(param.name), channels = param.channels, value = make_static_value(Color, param))

        elif isinstance(param, ImageParam):
            self._widget = GradioWidget(gr.Image, label = self._format_label(param.name), type = "numpy", image_mode = "RGBA" if param.channels == 4 else "RGB", value = make_static_value(np.ndarray, param))

        elif isinstance(param, ImageSourceParam):
            self._widget = ImageSourceEditor(label = self._format_label(param.name), channels = param.channels, value = make_static_value(ImageSource, param))

        elif isinstance(param, GradientParam):
            self._widget = GradientEditor(label = self._format_label(param.name), value = make_static_value(Gradient, param))

        elif isinstance(param, NoiseParam):
            self._widget = NoiseEditor(label = self._format_label(param.name), value = make_static_value(Noise, param))

        elif isinstance(param, PatternParam):
            self._widget = PatternEditor(label = self._format_label(param.name), value = make_static_value(Pattern, param))

        else:
            raise NotImplementedError

    @property
    def dependencies(self) -> Iterator[UIThing]:
        yield from self._widget.dependencies

    def read(self, data: ReadData) -> Any:
        return self._widget.read(data)

    def update(self, data: UpdateData) -> UpdateRequest:
        return self._widget.update(data)

    def setup_callback(self, callback: Callback) -> None:
        self._widget.setup_callback(callback)

File: ui/scripts/configurable_param_editor.js
import {Checkbox} from "../scripts/base/checkbox.js";
import {CodeArea} from "../scripts/base/code_area.js";
import {ColorPicker} from "../scripts/base/color_picker.js";
import {Dropdown} from "../scripts/base/dropdown.js";
import {ImageBox} from "../scripts/base/image_box.js";
import {NumberBox} from "../scripts/base/number_box.js";
import {Radio} from "../scripts/base/radio.js";
import {Slider} from "../scripts/base/slider.js";
import {TextArea} from "../scripts/base/text_area.js";
import {TextBox} from "../scripts/base/text_box.js";
import {VectorEditor} from "../scripts/base/vector_editor.js";
import {Signal} from "../scripts/core/signal.js";
import {Widget} from "../scripts/core/widget.js";
import {GradientEditor} from "../scripts/gradient_editor.js";
import {ImageSourceEditor} from "../scripts/image_source_editor.js";
import {NoiseEditor} from "../scripts/noise_editor.js";
import {PatternEditor} from "../scripts/pattern_editor.js";
import {ProcessingParamsEditor} from "../scripts/processing_params_editor.js";
import {VideoRendererEditor} from "../scripts/video_renderer_editor.js";

export class ConfigurableParamEditor extends Widget {
    constructor(definition) {
        super();

        this._editor = null;

        switch (definition.type) {
            case "bool": {
                this._editor = this.createChild(Checkbox, (e) => {
                    e.value = definition.value;
                });
            } break;

            case "int": {
                this._editor = this.createChild(definition.ui_type == "slider" ? Slider : NumberBox, (e) => {
                    e.minimum = definition.minimum ?? undefined;
                    e.maximum = definition.maximum ?? undefined;
                    e.step = definition.step ?? 1;
                    e.value = definition.default ?? e.minimum;
                });
            } break;

            case "float": {
                this._editor = this.createChild(definition.ui_type == "slider" ? Slider : NumberBox, (e) => {
                    e.minimum = definition.minimum ?? undefined;
                    e.maximum = definition.maximum ?? undefined;
                    e.step = definition.step ?? 0.1;
                    e.value = definition.default ?? e.minimum;
                });
            } break;

            case "str": {
                if (definition.ui_type == "menu" || definition.ui_type == "radio") {
                    this._editor = this.createChild(definition.ui_type == "radio" ? Radio : Dropdown, (e) => {
                        e.choices = definition.choices ?? {"": ""};
                        e.value = definition.default ?? null;
                    });
                } else {
                    this._editor = this.createChild(definition.ui_type == "code" ? CodeArea : definition.ui_type == "area" ? TextArea : TextBox, (e) => {
                        e.value = definition.default ?? "";
                    });

                }
            } break;

            case "pathlib.Path":{
                this._editor = this.createChild(TextBox, (e) => {
                    e.value = definition.default ?? "";
                });
            } break;

            case "numpy.ndarray": {
                this._editor = this.createChild(ImageBox, (e) => {
                    e.channels = definition.channels ?? 3;
                });
            } break;

            case "temporal.color.Color": {
                this._editor = this.createChild(ColorPicker, (e) => {
                    e.value = definition.default ?? {r: 0.0, g: 0.0, b: 0.0, a: 1.0};
                }, definition.channels ?? 3);
            } break;

            case "temporal.gradient.Gradient": {
                this._editor = this.createChild(GradientEditor, (e) => {
                    e.value = definition.default ?? {};
                });
            } break;

            case "temporal.image_source.ImageSource": {
                this._editor = this.createChild(ImageSourceEditor, (e) => {
                    e.channels = definition.channels ?? 3;
                });
            } break;

            case "temporal.noise.Noise": {
                this._editor = this.createChild(NoiseEditor, (e) => {
                    e.value = definition.default ?? {};
                });
            } break;

            case "temporal.pattern.Pattern": {
                this._editor = this.createChild(PatternEditor, (e) => {
                    e.value = definition.default ?? {};
                });
            } break;

            case "temporal.processing_params.ProcessingParams": {
                this._editor = this.createChild(ProcessingParamsEditor, (e) => {
                    e.value = definition.default ?? {};
                });
            } break;

            case "temporal.vector.IntVector": {
                this._editor = this.createChild(VectorEditor, (e) => {
                    e.minimum = definition.minimum ?? undefined;
                    e.maximum = definition.maximum ?? undefined;
                    e.step = definition.step ?? 1;
                    e.value = definition.default ?? {x: e.minimum, y: e.minimum};
                }, definition.ui_type == "slider" ? Slider : NumberBox, {
                    x: definition.axes?.[0] ?? "X",
                    y: definition.axes?.[1] ?? "Y",
                });
            } break;

            case "temporal.vector.FloatVector": {
                this._editor = this.createChild(VectorEditor, (e) => {
                    e.minimum = definition.minimum ?? undefined;
                    e.maximum = definition.maximum ?? undefined;
                    e.step = definition.step ?? 0.1;
                    e.value = definition.default ?? {x: e.minimum, y: e.minimum};
                }, definition.ui_type == "slider" ? Slider : NumberBox, {
                    x: definition.axes?.[0] ?? "X",
                    y: definition.axes?.[1] ?? "Y",
                });
            } break;

            case "temporal.video_renderer.VideoRenderer": {
                this._editor = this.createChild(VideoRendererEditor, (e) => {
                    e.value = definition.default ?? {};
                });
            } break;

            default: {
                console.log(`WARNING: Unhandled type ${definition.type}`);
            } break;
        }

        if (this._editor) {
            this._editor.onValueChange.connect((value) => {
                this.onValueChange.fire(value);
            });
        }

        this.onValueChange = new Signal();
    }

    get value() {
        return this._editor.value;
    }

    set value(value) {
        this._editor.value = value;
    }
}
customElements.define("configurable-param-editor", ConfigurableParamEditor);
